File: v6/prolog/labs/rel_spreading/probes/go/c2_collision.go
// C2 collision in Go: two embedded structs contributing the same field name.
// EXPECTED: the DECLARATION is accepted silently; the collision only surfaces
// at a use site, as an ambiguous selector.
package main

import "fmt"

type ARow struct {
	Shared int64
	OnlyA  int64
}

type BRow struct {
	Shared int64
	OnlyB  int64
}

// declared with a colliding promoted name: accepted
type Merged struct {
	ARow
	BRow
}

func main() {
	m := Merged{ARow{1, 2}, BRow{3, 4}}

	// the non-colliding promoted names work
	fmt.Println(m.OnlyA, m.OnlyB)

	// the colliding one is refused at the USE site only
	fmt.Println(m.Shared)
}

File: v6/prolog/labs/rel_spreading/probes/rust/c2_collision.rs
// C2 collision in Rust: what happens when two sources would contribute the
// same field name?

#[derive(Clone)]
struct ARow {
    shared: i64,
    only_a: i64,
}

#[derive(Clone)]
struct BRow {
    shared: i64,
    only_b: i64,
}

struct Merged {
    shared: i64,
    only_a: i64,
    only_b: i64,
}

fn main() {
    let a = ARow { shared: 1, only_a: 2 };
    let b = BRow { shared: 3, only_b: 4 };

    // two functional updates in one literal: refused
    let m = Merged { ..a, ..b };
    let _ = m.shared;

    // duplicate field name inside one struct declaration: refused
    let _ = Duplicated { shared: 1, shared: 2 };
}

struct Duplicated {
    shared: i64,
    shared: i64,
}
